tests/schema_update_test.go

```diff
@@ -1651,6 +1651,183 @@ func TestVCS_SQL_Review(t *testing.T) {
 	}
 }
 
+func TestBranchNameInVCSSetupAndUpdate(t *testing.T) {
+	externalID := "octocat/Hello-World"
+	repoFullPath := "octocat/Hello-World"
+
+	tests := []struct {
+		name               string
+		vcsProviderCreator fake.VCSProviderCreator
+		vcsType            vcs.Type
+		existedBranchList  []string
+		branchFilter       string
+		shouldFail         bool
+	}{
+		{
+			name:               "mainBranchWithGitLab",
+			vcsProviderCreator: fake.NewGitLab,
+			vcsType:            vcs.GitLabSelfHost,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "main",
+			shouldFail:   false,
+		},
+		{
+			name:               "mainBranchWithGitHub",
+			vcsProviderCreator: fake.NewGitHub,
+			vcsType:            vcs.GitHubCom,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "main",
+			shouldFail:   false,
+		},
+		{
+			name:               "customBranchWithGitLab",
+			vcsProviderCreator: fake.NewGitLab,
+			vcsType:            vcs.GitLabSelfHost,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "test/branch",
+			shouldFail:   false,
+		},
+		{
+			name:               "customBranchWithGitHub",
+			vcsProviderCreator: fake.NewGitHub,
+			vcsType:            vcs.GitHubCom,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "test/branch",
+			shouldFail:   false,
+		},
+		{
+			name:               "nonExistedBranchWithGitLab",
+			vcsProviderCreator: fake.NewGitLab,
+			vcsType:            vcs.GitLabSelfHost,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "non_existed_branch",
+			shouldFail:   true,
+		},
+		{
+			name:               "nonExistedBranchWithGitHub",
+			vcsProviderCreator: fake.NewGitHub,
+			vcsType:            vcs.GitHubCom,
+			existedBranchList: []string{
+				"main",
+				"test/branch",
+			},
+			branchFilter: "non_existed_branch",
+			shouldFail:   true,
+		},
+		{
+			name:               "wildcardBranchWithGitLab",
+			vcsProviderCreator: fake.NewGitLab,
+			vcsType:            vcs.GitLabSelfHost,
+			existedBranchList: []string{
+				"main",
+			},
+			branchFilter: "main*",
+			shouldFail:   false,
+		},
+		{
+			name:               "wildcardBranchWithGitHub",
+			vcsProviderCreator: fake.NewGitHub,
+			vcsType:            vcs.GitHubCom,
+			existedBranchList: []string{
+				"main",
+			},
+			branchFilter: "main*",
+			shouldFail:   false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			port := getTestPort(t.Name())
+			a := require.New(t)
+			ctx := context.Background()
+			ctl := &controller{}
+			err := ctl.StartServer(ctx, t.TempDir(), test.vcsProviderCreator, port)
```

Comment:
Can we not start a new server for every sub-test? Can we just create new project instead?

Reply:
Since we are testing on both GitLab and GitHub. We can start a new server and create a new project for each of them.